### PR TITLE
Add Model Surgery HOWTO

### PR DESCRIPTION
**Pull request by marcvanzee**
This HOWTO is from a Colab by @levskaya.

### CODE REVIEW

File: docs/howtos/model_surgery.rst [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  import jax
+  from jax import lax, random, numpy as jnp
+  import flax
+  from flax import optim, traverse_util
+
+  from flax import linen as nn
+  from flax.core import unfreeze, freeze
+
+We will show how to get a flat dict of all the tensors, and then go back to a 
+nested, frozen dict. We show this for both Flax modules and optimizers.
```
Comment (8bitmp3) [inline docs/howtos/model_surgery.rst:17]: Suggestion:

- To avoid the "We ... show.... We show" repetition, maybe you can phrase it as follows:

```suggestion
We will show how to get a flat dict of all the tensors, and then go back to a 
nested, frozen dict. This will be demonstrated for both Flax modules and optimizers.
```

File: docs/howtos/model_surgery.rst [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+Surgery with Flax Modules
+--------------------------------
+
+Let's first make a little model for our demo.
```
Comment (8bitmp3) [inline docs/howtos/model_surgery.rst:22]: Suggestion:

- For those use who have knowledge of numerical computing (NumPy) but not as familiar with NNs, I propose to maybe rephrase this like so:

```suggestion
Let's create a small convolutional neural network model for our demo.
```

File: docs/howtos/model_surgery.rst [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+          kernel: (3136, 256),
+      },
+      Dense_1: {
+          bias: (10,),
+          kernel: (256, 10),
+      },
+  })
+
+
+Let's get a flat dict for doing model surgery.
```
Comment (8bitmp3) [inline docs/howtos/model_surgery.rst:74]: Suggestion:

- To avoid the "Let's... Let's" repetition, maybe this would sound better:

```suggestion
Next, get a flat dict for doing model surgery as follows:
```

File: docs/howtos/model_surgery.rst [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+      Dense_1: {
+          bias: (10,),
+          kernel: (256, 10),
+      },
+  })
+
+Surgey with Optimizers
+--------------------------------
+
+If you're loading from a flax optimizer, all of the variables live in
```
Comment (8bitmp3) [inline docs/howtos/model_surgery.rst:131]: Suggestion: nit

```suggestion
If you're loading from a Flax optimizer, all of the variables live in
```

**Review by 8bitmp3 — COMMENTED** [review]
This is a great HOWTO, @marcvanzee. I have proposed a few minor suggestions to improve it a little but feel free to ignore them. Thank you

File: docs/howtos/model_surgery.rst [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+          bias: (256,),
+          kernel: (3136, 256),
+      },
+      Dense_1: {
+          bias: (10,),
+          kernel: (256, 10),
+      },
+  })
+
+Surgey with Optimizers
```
Comment (andsteing) [inline docs/howtos/model_surgery.rst:128]: Surgery

File: docs/howtos/model_surgery.rst (hunk 4 above)
Comment (andsteing) [inline docs/howtos/model_surgery.rst:131]: it's only the variables that should be optimized by the optimizer, usually the `params` collection.

File: docs/howtos/model_surgery.rst [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+nested, frozen dict. We show this for both Flax modules and optimizers.
+
+Surgery with Flax Modules
+--------------------------------
+
+Let's first make a little model for our demo.
+
+.. testcode::
+
+  class CNN(nn.Module):
```
Comment (andsteing) [inline docs/howtos/model_surgery.rst:26]: Why this particular structure?
Did you consider making it shorter not to distract from the more important code?
(Or maybe adding another level of modules if you think that's insightful)

**Review by andsteing — APPROVED** [review]

**Comment by marcvanzee** [comment]
> This is a great HOWTO, @marcvanzee. I have proposed a few minor suggestions to improve it a little but feel free to ignore them. Thank you

Thanks a lot for your comments! I incorporated them.

File: docs/howtos/model_surgery.rst (hunk 6 above)
Comment (marcvanzee) [reply docs/howtos/model_surgery.rst:26]: It's the default CNN that's used in many of our HOWTOs. We could change it, but I actually think it is quite nice since it simply yet realistic.

**Review by levskaya — APPROVED** [review]

**Comment by 8bitmp3** [comment]
Thanks @marcvanzee This is awesome
